css-inline/src/html/serializer.rs

```
@@ -114,6 +127,22 @@ impl<'a> Sink<'a> {
 
                 serializer.start_elem(&element.name, &element.attributes, style_node_id)?;
 
+                // TODO this part is the one that I don't like the most
```

I am not 100% sure about this one, even though it is a sound approach.

The first thought that I have is that the least surprising thing would be to keep `@` rules in their original places, instead of pushing everything into a single node. This won't work for external stylesheets, so we need to store them somewhere anyway.

On the other hand, this approach is better for performance, and probably it would not be super surprising for the user anyway if we put everything under one `<style>` tag.

So, in the end, I am leaning towards this approach.

A few other notes:

1. We can reduce the performance hit of checking each node name by having two versions of `serialize` - the first one will work as before this PR. The new one will check for `head` and once it is found, it will switch to the other version of `serialize` that won't check the tag name. At the top level we check if `at_rules` are present and then call one of these two versions - I think it should minimize the impact as `head` is usually in the very beginning of the nodes vector.
2. Are there cases when `head` is not present? Maybe in HTML fragments? see `inline_fragment`

---

btw, if the user chose to keep `style` tags - I think we can only extract `@` rule from `link`, as the old `style` tags will be around anyway

---

That's something I had no time to look closer into yet

---

Given the current benchmark results, the performance hit is negligible, so we can skip this. For fragments, likely we can't really do it, as there is no regular HTML structure generally. Maybe we can return an error when the user sets an option to keep at rules, but uses `inline_fragment`?